src/hotspot/share/opto/castnode.cpp

```diff
@@ -371,6 +371,29 @@ void CastIINode::dump_spec(outputStream* st) const {
 }
 #endif
 
+Node *CastLLNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  Node* progress = ConstraintCastNode::Ideal(phase, can_reshape);
+  if (progress != NULL) {
+    return progress;
+  }
+  // transform (CastLL (ConvI2L ..)) into (ConvI2L (CastII ..)) if the type of the CastLL is narrower than the type of
+  // the ConvI2L.
+  Node* in1 = in(1);
+  if (in1 != NULL && in1->Opcode() == Op_ConvI2L) {
+    const Type* t = Value(phase);
+    const Type* t_in = phase->type(in1);
+    if (t != Type::TOP && t_in != Type::TOP && t != t_in) {
```

Comment:
`t != t_in` does not mean that type is narrower in general case. I think we need to check ranges (types meet?).

Reply:
There is no specialized `CastLLNode::Value()` and `ConstraintCastNode` only calls `filter_speculative()` which do call `join()`. May be it is indeed enough. Yes, would be nice to have an assert to make sure we got it right.

Reply:
The new commit adds an assert